File: backend/app/api/monitors.py
"""Monitors API — inbox filters + URL diff monitors (APP_STORAGE, no Kernel events)."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.product import inbox_monitors as inbox_mon
from app.product import url_monitors as url_mon

router = APIRouter(tags=["monitors"])


# ── Inbox filters ────────────────────────────────────────────────────────────


class InboxFilterCreate(BaseModel):
    name: str = ""
    sender_contains: str = ""
    subject_contains: str = ""
    enabled: bool = True


class InboxFilterUpdate(BaseModel):
    name: str | None = None
    sender_contains: str | None = None
    subject_contains: str | None = None
    enabled: bool | None = None


@router.get("/inbox-filters")
async def list_inbox_filters():
    return {"filters": inbox_mon.list_inbox_filters()}


@router.post("/inbox-filters")
async def create_inbox_filter(body: InboxFilterCreate):
    try:
        row = inbox_mon.create_inbox_filter(
            name=body.name,
            sender_contains=body.sender_contains,
            subject_contains=body.subject_contains,
            enabled=body.enabled,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return row


@router.patch("/inbox-filters/{filter_id}")
async def update_inbox_filter(filter_id: str, body: InboxFilterUpdate):
    updates = body.model_dump(exclude_unset=True)
    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")
    try:
        return inbox_mon.update_inbox_filter(filter_id, **updates)
    except KeyError:
        raise HTTPException(status_code=404, detail="Filter not found") from None
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.delete("/inbox-filters/{filter_id}")
async def delete_inbox_filter(filter_id: str):
    try:
        inbox_mon.delete_inbox_filter(filter_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="Filter not found") from None
    return {"status": "ok"}


# ── URL monitors ─────────────────────────────────────────────────────────────


class UrlMonitorCreate(BaseModel):
    name: str = ""
    url: str = ""
    enabled: bool = True
    check_interval_minutes: int = Field(
        default=url_mon.DEFAULT_INTERVAL_MINUTES,
        ge=url_mon.MIN_INTERVAL_MINUTES,
        le=url_mon.MAX_INTERVAL_MINUTES,
    )


class UrlMonitorUpdate(BaseModel):
    name: str | None = None
    url: str | None = None
    enabled: bool | None = None
    check_interval_minutes: int | None = Field(
        default=None,
        ge=url_mon.MIN_INTERVAL_MINUTES,
        le=url_mon.MAX_INTERVAL_MINUTES,
    )


@router.get("/url-monitors")
async def list_url_monitors():
    return {"monitors": url_mon.list_url_monitors()}


@router.post("/url-monitors")
async def create_url_monitor(body: UrlMonitorCreate):
    try:
        return url_mon.create_url_monitor(
            name=body.name,
            url=body.url,
            enabled=body.enabled,
            check_interval_minutes=body.check_interval_minutes,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/url-monitors/check")
async def check_url_monitors(force: bool = True):
    """Run due (or all, when force=True) URL monitors once — dogfood / manual.

    Caps forced runs at 5 pages so the HTTP request cannot hang for minutes.
    """
    max_checks = 5 if force else None
    notified = await url_mon.evaluate_url_monitors(force=force, max_checks=max_checks)
    return {"notified": notified, "force": force, "max_checks": max_checks}


@router.patch("/url-monitors/{monitor_id}")
async def update_url_monitor(monitor_id: str, body: UrlMonitorUpdate):
    updates = body.model_dump(exclude_unset=True)
    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")
    try:
        return url_mon.update_url_monitor(monitor_id, **updates)
    except KeyError:
        raise HTTPException(status_code=404, detail="URL monitor not found") from None
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.delete("/url-monitors/{monitor_id}")
async def delete_url_monitor(monitor_id: str):
    try:
        url_mon.delete_url_monitor(monitor_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="URL monitor not found") from None
    return {"status": "ok"}

File: frontend/src/api/monitors.ts
/** Monitors API — inbox filters + URL diff monitors. */

import { API_BASE, request } from "./core";

export interface InboxFilter {
  id: string;
  enabled: boolean;
  name: string;
  sender_contains: string;
  subject_contains: string;
  created_at?: string;
}

export interface UrlMonitor {
  id: string;
  enabled: boolean;
  name: string;
  url: string;
  check_interval_minutes: number;
  created_at?: string;
  last_hash?: string | null;
  last_title?: string | null;
  last_checked_at?: string | null;
  last_error?: string | null;
}

export async function listInboxFilters(): Promise<InboxFilter[]> {
  const data = await request<{ filters: InboxFilter[] }>(`${API_BASE}/monitors/inbox-filters`);
  return data.filters ?? [];
}

export async function createInboxFilter(body: {
  name: string;
  sender_contains?: string;
  subject_contains?: string;
  enabled?: boolean;
}): Promise<InboxFilter> {
  return request<InboxFilter>(`${API_BASE}/monitors/inbox-filters`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateInboxFilter(
  filterId: string,
  body: Partial<Pick<InboxFilter, "name" | "sender_contains" | "subject_contains" | "enabled">>,
): Promise<InboxFilter> {
  return request<InboxFilter>(`${API_BASE}/monitors/inbox-filters/${filterId}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteInboxFilter(filterId: string): Promise<void> {
  await request(`${API_BASE}/monitors/inbox-filters/${filterId}`, { method: "DELETE" });
}

export async function listUrlMonitors(): Promise<UrlMonitor[]> {
  const data = await request<{ monitors: UrlMonitor[] }>(`${API_BASE}/monitors/url-monitors`);
  return data.monitors ?? [];
}

export async function createUrlMonitor(body: {
  name: string;
  url: string;
  enabled?: boolean;
  check_interval_minutes?: number;
}): Promise<UrlMonitor> {
  return request<UrlMonitor>(`${API_BASE}/monitors/url-monitors`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateUrlMonitor(
  monitorId: string,
  body: Partial<Pick<UrlMonitor, "name" | "url" | "enabled" | "check_interval_minutes">>,
): Promise<UrlMonitor> {
  return request<UrlMonitor>(`${API_BASE}/monitors/url-monitors/${monitorId}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteUrlMonitor(monitorId: string): Promise<void> {
  await request(`${API_BASE}/monitors/url-monitors/${monitorId}`, { method: "DELETE" });
}

export async function checkUrlMonitors(force = true): Promise<{ notified: number }> {
  return request<{ notified: number; force: boolean }>(
    `${API_BASE}/monitors/url-monitors/check?force=${force ? "true" : "false"}`,
    { method: "POST" },
  );
}
